### PR TITLE
feat: support reference loops

### LINKED ISSUE #172: Circular References Support
**Describe the bug**
circular dependency will cause StackOverflowException

**To Reproduce**
1. Declare a mapper as an interface
2. Add a mapping method `CarDto CarToDto(Car car);`
3. Add a mapping for "Seats" wich refernce to Car via ParentCar property
4. Add property to Car called "Primary Seat" which is of type Seat
5. Assign everything

```
Stack overflow.
Repeat 3212 times:
--------------------------------
   at ConsoleApp59.CarMapper.CarToCarDto(ConsoleApp59.Car)
   at ConsoleApp59.CarMapper.MapToSeatDto(ConsoleApp59.Seat)
   at ConsoleApp59.CarMapper.MapToSeatDtoArray(ConsoleApp59.Seat[])
--------------------------------
   at ConsoleApp59.CarMapper.CarToCarDto(ConsoleApp59.Car)
   at Program.<Main>$(System.String[])
```

**Expected behavior**
I expect mapper is smart enough and when it is tasked with mapping instance which it delt before with (reference equal) it would return already newed instance. 

**Code snippets**
checkout this repo https://github.com/vmachacek/mapperly-circular-references

**Environment (please complete the following information):**
 - Mapperly Version: 2.5.0
 - .NET Version: NET 6.0.401
 - Target Framework: net6.0
 - OS: Windows

### PR DESCRIPTION
**Pull request by latonz**
Closes #172

### CODE REVIEW

File: src/Riok.Mapperly/Descriptors/Mappings/ITypeMapping.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Riok.Mapperly.Descriptors.Mappings;
+
+/// <summary>
+/// Represents a mapping to map from one type to another.
```
Comment (CommonGuy) [inline src/Riok.Mapperly/Descriptors/Mappings/ITypeMapping.cs:7]: I think "to map" can be omitted here

File: src/Riok.Mapperly.Abstractions/MapperAttribute.cs [hunk 2]
```diff
@@ -60,4 +62,12 @@ public sealed class MapperAttribute : Attribute
     /// </example>
     /// </summary>
     public MappingConversionType EnabledConversions { get; set; } = MappingConversionType.All;
+
+    /// <summary>
+    /// Enables the reference handling feature.
+    /// Disabled by default for performance reasons.
+    /// When enabled, a <see cref="IReferenceHandler"/> instance is passed through the mapping methods
```
Comment (CommonGuy) [inline src/Riok.Mapperly.Abstractions/MapperAttribute.cs:69]: Small typo, should be
```
an <see cref="IReferenceHandler"/>
```

Comment (latonz) [reply src/Riok.Mapperly.Abstractions/MapperAttribute.cs:69]: Improved.

File: src/Riok.Mapperly/Descriptors/Mappings/ITypeMapping.cs (hunk 1 above)
Comment (latonz) [reply src/Riok.Mapperly/Descriptors/Mappings/ITypeMapping.cs:7]: Improved.

**Review by CommonGuy — DISMISSED** [review]

**Review by CommonGuy — APPROVED** [review]

**Pull request merged**